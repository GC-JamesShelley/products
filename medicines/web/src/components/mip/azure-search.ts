const azureApiVersion = process.env.AZURE_API_VERSION;
const azureIndex = process.env.AZURE_INDEX;
const azureKey = process.env.AZURE_KEY;
const azureService = process.env.AZURE_SERVICE;
const azureWordFuzziness = process.env.AZURE_WORD_FUZZINESS;

enum DocType {
  PilLabel,
  PilLabelAndLeaflet,
  PilLeaflet,
  Spc,
}

export interface IAzureSearchResult {
  '@search.score': number;
  '@search.highlights': { content: string[] };
  content: string;
  doc_type: DocType;
  metadata_storage_path: string;
}

const buildFuzzyQuery = (query: string): string =>
  query
    .split(' ')
    .map(word => `${word}~${azureWordFuzziness}`)
    .join(' ');

const buildAzureSearchUrl = (query: string): string => {
  const url = new URL(
    `https://${azureService}.search.windows.net/indexes/${azureIndex}/docs`,
  );

  url.searchParams.append('api-key', azureKey as string);
  url.searchParams.append('api-version', azureApiVersion as string);
  url.searchParams.append('highlight', 'content');
  url.searchParams.append('queryType', 'full');
  url.searchParams.append('search', query);

  return url.toString();
};

const getJson = async (url: string): Promise<any> => {
  const resp: Response = await fetch(url, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (resp.ok) {
    return resp.json();
  }
};

export const azureSearch = async (
  query: string,
): Promise<IAzureSearchResult[]> => {
  const body = await getJson(buildAzureSearchUrl(buildFuzzyQuery(query)));

  return body.value;
};
