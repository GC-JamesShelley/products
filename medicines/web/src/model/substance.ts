export interface IProduct {
  name: string;
  count?: number;
}

export interface ISubstance extends IProduct {
  products?: IProduct[];
}

export function isSubstance(obj: any): obj is ISubstance {
  return obj.products !== undefined && obj.count !== undefined;
}

export function isIndex(obj: any): obj is ISubstance {
  return obj.products === undefined && obj.count === undefined;
}

export interface IDocument {
  activeSubstances: string[];
  context: string;
  created: string;
  docType: string;
  fileSize: string;
  name: string;
  product: string;
  url: string;
}
