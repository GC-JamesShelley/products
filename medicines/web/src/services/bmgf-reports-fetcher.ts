export const getReportList = async (
  containerUrl: string,
): Promise<string[]> => {
  return fetch(`${containerUrl}/upload-index.txt`, {
    method: 'GET',
    headers: {
      'Content-Type': 'text/text',
    },
  })
    .then((response) => response.text())
    .then((text) => {
      return text.split('\n').filter((report) => report.length > 0);
    });
};

const getReportUrl = async (
  reportToGet: string,
  containerUrl: string,
): Promise<any> => {
  return getReportList(containerUrl)
    .then((reports) => {
      return reports.find((report) => report.startsWith(reportToGet));
    })
    .then((report) => `${containerUrl}/${report}`);
};

export const getReportContent = async (
  report: string,
  containerUrl: string,
): Promise<string> => {
  return getReportUrl(report, containerUrl).then((reportUrl) => {
    return fetch(reportUrl, {
      method: 'GET',
      headers: {
        'Content-Type': 'text/html',
      },
    }).then((response) => response.text());
  });
};
