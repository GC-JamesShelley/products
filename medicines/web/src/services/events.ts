import TagManager from 'react-gtm-module';

const pushToDataLayer = (dataLayer: any) => {
  TagManager.dataLayer({
    dataLayer,
  });
};

export default {
  searchForProductsMatchingKeywords: (searchTerm: string, pageNo: number) => {
    pushToDataLayer({
      event: 'search',
      searchTerm,
      pageNo,
    });
  },
  viewProductsForSubstance: (substance: string) => {
    pushToDataLayer({ event: 'substance', substance });
  },
  viewSubstancesStartingWith: (letter: string) => {
    pushToDataLayer({ event: 'drugIndex', letter });
  },
  viewPage: (pageName: string) => {
    pushToDataLayer({ event: pageName });
  },
};
