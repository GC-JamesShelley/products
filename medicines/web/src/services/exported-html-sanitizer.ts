const updateImageTag = (imageNode, prefix) => {
  for (let i = 0; i < imageNode.attrs.length; i++) {
    if (imageNode.attrs[i].name === 'src') {
      const imageName = imageNode.attrs[i].value.split('/').pop();
      imageNode.attrs[i].value = encodeURI(`${prefix}${imageName}`);
    } else if (imageNode.attrs[i].name === 'v:shapes') {
      imageNode.attrs.splice(i, 1);
      i--;
    }
  }
  return imageNode;
};

const updateAnchorNameToId = (node) => {
  for (const attribute of node.attrs) {
    if (attribute.name === 'name') {
      attribute.name = 'id';
      return node;
    }
  }
  return node;
};

const removeUnwantedTableAttributes = (node) => {
  if (!node.attrs) {
    return node;
  }

  const unwantedAttributes = [
    'style',
    'v:shapes',
    'cellspacing',
    'cellpadding',
    'border',
    'width',
    'valign',
    'class',
  ];

  for (let i = 0; i < node.attrs.length; i++) {
    if (unwantedAttributes.includes(node.attrs[i].name)) {
      node.attrs.splice(i, 1);
      i--;
    }
  }

  return node;
};

const removeUnwantedAttributes = (node) => {
  if (!node.attrs) {
    return node;
  }

  const unwantedAttributes = ['style', 'align', 'class'];
  for (let i = 0; i < node.attrs.length; i++) {
    if (unwantedAttributes.includes(node.attrs[i].name)) {
      node.attrs.splice(i, 1);
      i--;
    }
  }
  return node;
};

const tagShouldBeRemoved = (tagName) => {
  return ['h1', 'o:p', 'w:sdt'].includes(tagName);
};

export const recurseNodes = (node, prefix) => {
  if (tagShouldBeRemoved(node.tagName)) {
    return;
  }
  if (node.tagName === 'img') {
    node = updateImageTag(node, prefix);
  } else if (node.tagName === 'td' || node.tagName === 'table') {
    node = removeUnwantedTableAttributes(node);
  } else if (node.tagName === 'a') {
    node = updateAnchorNameToId(node);
  } else {
    node = removeUnwantedAttributes(node);
  }

  for (let i = 0; i < node.childNodes?.length ?? 0; i++) {
    const returnedNode = recurseNodes(node.childNodes[i], prefix);
    if (returnedNode) {
      node.childNodes[i] = returnedNode;
    } else {
      node.childNodes.splice(i, 1);
      i--;
    }
  }
  return node;
};

export const cleanUpHtml = (htmlBody, assetPrefix) => {
  return recurseNodes(htmlBody, assetPrefix);
};

export const getHtmlBody = (htmlDoc) => {
  const html = htmlDoc.childNodes[0];
  for (const node of html.childNodes) {
    if (node.tagName === 'body') {
      return node;
    }
  }
};
