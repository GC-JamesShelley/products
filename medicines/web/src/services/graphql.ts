interface IGraphQLResponse<T> {
  data: T;
  // TODO: add proper type for errors field
}

const graphqlUrl = process.env.GRAPHQL_URL as string;

export async function graphqlRequest<T, V extends {}>({
  query,
  variables,
}: {
  query: string;
  variables: V;
}): Promise<IGraphQLResponse<T>> {
  const response = await fetch(graphqlUrl, {
    method: 'POST',
    mode: 'cors',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query,
      variables,
    }),
  });

  if (!response.ok) {
    throw new Error(`Error response code from GraphQL API: ${response.status}`);
  }

  return response.json();
}
