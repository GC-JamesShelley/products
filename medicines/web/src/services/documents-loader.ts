import DataLoader from 'dataloader';
import { IDocument } from '../model/substance';

const graphQlUrl = process.env.GRAPHQL_URL as string;

const fetchFromGraphQl = (query: string): Promise<Response> => {
  return fetch(graphQlUrl, {
    method: 'POST',
    mode: 'cors',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ query, variables: null }),
  });
};

const getDocumentsForProduct = async (productName: any) => {
  const query = `{ product(name: "${productName}") { documents {       productName
    activeSubstances
    title
    highlights
    created
    docType
    fileBytes
    name
    url } } }`;
  const response = await fetchFromGraphQl(query);

  return (await response.json()).data.product.documents;
};

export const documents = new DataLoader<string, IDocument[]>(
  async productNames => {
    return Promise.all(productNames.map(getDocumentsForProduct));
  },
);
