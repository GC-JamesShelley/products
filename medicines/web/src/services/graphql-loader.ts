import DataLoader from 'dataloader';
import { IProduct } from '../model/substance';

const graphQlUrl = process.env.GRAPHQL_URL as string;

const getProductsForSubstance = async (substanceName: string) => {
  const query = `
query ($substanceName: String) {
  substance(name: $substanceName) {
    products {
      name
      documentCount
    }
  }
}`;

  const variables = { substanceName };

  const response = await fetch(graphQlUrl, {
    method: 'POST',
    mode: 'cors',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query,
      variables,
    }),
  });

  return (await response.json()).data.substance.products;
};

const products = new DataLoader<string, IProduct[]>(async substanceNames => {
  return Promise.all(substanceNames.map(getProductsForSubstance));
});

export default {
  products,
};
