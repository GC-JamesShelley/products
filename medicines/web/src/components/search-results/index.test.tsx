import { shallow } from 'enzyme';
import React from 'react';
import { RerouteType } from '../../model/rerouteType';
import { DocType } from '../../services/azure-search';
import SearchResults from './index';

const drugsMock = [
  {
    activeSubstances: ['tea', 'coffee'],
    context: 'string',
    created: 'string',
    docType: 'string',
    fileSize: 'string',
    name: 'string',
    product: 'string',
    url: 'string',
  },
  {
    activeSubstances: ['tea', 'coffee'],
    context: 'string',
    created: 'string',
    docType: 'string',
    fileSize: 'string',
    name: 'string',
    product: 'string',
    url: 'string',
  },
];

const noFeedback = () => undefined;
const updateDocType = (d: DocType[]) => undefined;

describe(SearchResults, () => {
  it('should render', () => {
    const component = shallow(
      <SearchResults
        drugs={drugsMock}
        page={1}
        pageSize={20}
        resultCount={200}
        searchTerm={'Tea'}
        showingResultsForTerm={'Tea'}
        disclaimerAgree
        docTypes={[]}
        updateDocTypes={updateDocType}
        rerouteType={RerouteType.CheckboxSelected}
        handlePageChange={noFeedback}
        isLoading={false}
      />,
    );
    expect(component).toMatchSnapshot();
  });
});
