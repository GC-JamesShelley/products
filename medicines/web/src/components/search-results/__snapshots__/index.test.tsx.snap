// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults should render 1`] = `
<Fragment>
  <styled.div>
    <div>
      <h1
        className="title"
      >
        Showing results for Tea
      </h1>
      <p
        className="no-of-results"
      >
        1 to 2 of 200
      </p>
      <p
        className="ema-message"
      >
        If the product information you are seeking does not appear below, it is possible that the product holds a European licence and its information may be available at the 
        <a
          href="https://www.ema.europa.eu/en"
          target="_new"
        >
          European Medicines Agency
        </a>
         website.
      </p>
      <p>
        Before a medicine can be sold in the UK, a number of licences are essential. Products with a UK marketing authorisation have a licence number in the format ‘PL 12345/0001’. The first 2 characters are always the letters ‘PL’, and this can be found on the packaging of the product.
      </p>
      <p>
        You can identify the product in the list below using the PL number.
      </p>
      <p>
        The information about a medicine will be updated when new evidence becomes available. This may mean that there are differences between the information in the pack and the information here. The most up-to-date information will be available on this site.
      </p>
    </div>
    <div
      className="row"
    >
      <div
        className="column filter"
      >
        <SearchFilter
          currentlyEnabledDocTypes={Array []}
          toggleDocType={[Function]}
        />
      </div>
      <div
        className="column results"
      >
        <dl>
          <article
            key="0"
          >
            <dt
              className="left"
            >
              <p
                className="icon"
              >
                STRING
              </p>
            </dt>
            <dd
              className="right"
            >
              <a
                className="doc-type-string"
                href="string"
              >
                <p
                  className="title"
                >
                  string
                </p>
                <p
                  className="subtitle"
                >
                  string
                </p>
              </a>
              <p
                className="metadata"
              >
                File size: 
                string
                 KB
              </p>
              <p
                className="metadata"
              >
                Active substances:
                 
                Tea, Coffee
              </p>
              <p
                className="context"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "string",
                  }
                }
              />
            </dd>
          </article>
          <article
            key="1"
          >
            <dt
              className="left"
            >
              <p
                className="icon"
              >
                STRING
              </p>
            </dt>
            <dd
              className="right"
            >
              <a
                className="doc-type-string"
                href="string"
              >
                <p
                  className="title"
                >
                  string
                </p>
                <p
                  className="subtitle"
                >
                  string
                </p>
              </a>
              <p
                className="metadata"
              >
                File size: 
                string
                 KB
              </p>
              <p
                className="metadata"
              >
                Active substances:
                 
                Tea, Coffee
              </p>
              <p
                className="context"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "string",
                  }
                }
              />
            </dd>
          </article>
        </dl>
      </div>
    </div>
  </styled.div>
  <Pagination
    currentPage={1}
    pageSize={20}
    resultCount={200}
    searchTerm="Tea"
  />
</Fragment>
`;
